### DISCUSSION TITLE
How to find rendered URL

### DISCUSSION THREAD

**Discussion by f0o** [sindresorhus/got#1809] [Q&A] [answered]
#### What would you like to discuss?
How can I find out the rendered URL with all searchParams etc?

I'm having some issues where I expect the searchParams being wrongly encoded and would like to verify that by seeing what's actually being sent...

...

#### Checklist

- [X] I have read the documentation.

**Comment by sindresorhus** [comment, marked as answer]
you can get the rendered URL in a few ways:

**after the request completes:**
```js
const response = await got('https://example.com', {
	searchParams: {foo: 'bar', baz: 'qux'}
});

console.log(response.requestUrl.toString());
// => https://example.com?foo=bar&baz=qux

console.log(response.url);
// => final URL as string (after redirects)
```

**before the request is sent (to debug encoding):**
```js
await got('https://example.com', {
	searchParams: {foo: 'bar baz'},
	hooks: {
		beforeRequest: [
			options => {
				console.log(options.url.toString());
				// => https://example.com?foo=bar+baz
			}
		]
	}
});
```

the `beforeRequest` hook is the best way to see exactly whats being sent, including how searchParams are encoded.
